node/src/frost_sign.rs

```diff
@@ -0,0 +1,576 @@
+    fn handle_ready_to_send_commitment_state(&mut self) -> anyhow::Result<Action<FrostOutput>> {
+        let coordinator = if let Some(coordinator) = self.coordinator {
+            coordinator
+        } else {
+            return Ok(Action::Wait);
+        };
+
+        self.state = ParticipantState::ReadyToSendSigningShare;
+
+        let data = self.my_commitment.serialize()?;
+        let data = ParticipantMessage::Commitment(data).into_data()?;
+        Ok(Action::SendPrivate(coordinator, data))
+    }
+
+    fn handle_ready_to_send_signing_share_state(&mut self) -> anyhow::Result<Action<FrostOutput>> {
+        let coordinator = if let Some(coordinator) = self.coordinator {
+            coordinator
+        } else {
+            return Ok(Action::Wait);
+        };
+
+        let signing_package = if let Some(signing_package) = self.signing_package.as_ref() {
+            signing_package
+        } else {
+            return Ok(Action::Wait);
+        };
+
+        let signing_package =
+            frost_ed25519::SigningPackage::deserialize(signing_package.as_slice())?;
+
+        if signing_package
+            .signing_commitment(self.key_package.identifier())
+            .is_none()
+        {
+            // We were not fast enough to get included into first `threshold` participants.
+            self.state = ParticipantState::Finished;
+            return self.poke_state_machine();
+        }
+
+        let signature_share = frost_ed25519::round2::sign(
+            &signing_package,
+            &self.my_signing_nonce,
+            &self.key_package,
+        )?;
+
+        self.state = ParticipantState::Finished;
+        let data = ParticipantMessage::SignatureShare(signature_share.serialize());
+        let data = data.into_data()?;
+        Ok(Action::SendPrivate(coordinator, data))
+    }
+
+    fn poke_state_machine(&mut self) -> anyhow::Result<Action<FrostOutput>> {
+        match self.state {
+            ParticipantState::ReadyToSendCommitment => self.handle_ready_to_send_commitment_state(),
+            ParticipantState::ReadyToSendSigningShare => {
+                self.handle_ready_to_send_signing_share_state()
+            }
+            ParticipantState::Finished => Ok(Action::Return(FrostOutput::Participant)),
+        }
+    }
+    fn handle_message(&mut self, from: Participant, msg: CoordinatorMessage) {
+        match msg {
+            CoordinatorMessage::FirstRoundInit => {
+                if self.coordinator.is_none() {
+                    self.coordinator = Some(from)
+                }
+            }
+            CoordinatorMessage::SigningPackage(signing_package) => {
+                if Some(from) == self.coordinator {
+                    self.signing_package = Some(signing_package);
+                } else {
+                    self.state = ParticipantState::Finished
+                }
+            }
+        }
+    }
+}
+
+impl Protocol for FrostSignProtocolParticipant {
```

Comment:
Maybe add a high level summary of what we're doing here, i.e. adapting the frost protocol to the API of the cait-sith protocol.